### ISSUE TITLE
[csrng] input seed length shall be larger than 384 bits

### ISSUE THREAD

**Issue by vsukhoml** [lowRISC/opentitan#22157]
### Description

[NIST 800-90A](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf) states:
1. 8.6.3: The entropy input shall have entropy that is equal to or greater than the security strength of the
instantiation. Additional entropy may be provided in the nonce or the optional personalization
string during instantiation, or in the additional input during reseeding and generation, but this is
not required and does not increase the “official” security strength of the DRBG instantiation that
is recorded in the internal state.
For us it means at least 256 bits of full entropy.

2.  8.6.7: A nonce may be required in the construction of a seed during instantiation in order to provide a
security cushion to block certain attacks. The nonce shall be either:  A value with at least (security_strength/2) bits of entropy, or A value that is expected to repeat no more often than a (security_strength/2)-bit random
string would be expected to repeat. 

3. 8.7.1: A personalization string is an optional (but recommended) input to the instantiate function and is
used to derive the seed (see Section 8.6.1). The personalization string may be obtained from
inside or outside a cryptographic module, and may be an empty string. Note that a DRBG does
not rely on a personalization string to provide entropy, even though entropy could be provided in
the personalization string, and knowledge of the personalization string by an adversary does not
degrade the security strength of a DRBG instantiation, as long as the entropy input is unknown. 

So the problem is that if nonce is used, then there is no space for personalization string. Even if nonce is not used, personalization string is limited to 4 words, which would mean truncation of hashes.

It would be highly desirable to increase total seed length to at least 512 or, better, 768 bits to support variety of use cases.

However, for CTR_DRBG in 10.2, seed length is limited to 384 bits if no derivation is used. This bring a question where derivation better to happen in software or hardware. 10.3.2 describes quite unique Derivation Function Using a Block Cipher Algorithm (Block_Cipher_df), which runs AES-CBC for CTR_DRBG.

Pros for hardware: simpler software interface (same register level), ability to sideload values from key manager to mix in (issue #22117 ). Cons: more complex HW.

Pros for software: no changes in HW. Cons no sideloading from key manager.

**Comment by johannheyszl** [comment]
The root issues is described in https://github.com/lowRISC/opentitan/issues/21936. Need for changes to CSRNG will depend on its resolution.

**Comment by moidx** [comment]
Moving to M4 for triaging purposes.

@vogelpi: Changing the user provided input length is most likely not possible at this time as this will have an area impact, as well as, it will require significant re-architecture of the CSRNG block. Currently, the additional input in the CSRNG command interface can be used to add a 12 words personalization string in the `generate` command. 

@vsukhoml and @ballifatih to discuss if the generate command additional input may be suitable for the use cases in question.

**Comment by andreaskurth** [comment]
By @johannheyszl: This was created before we decided to use a software DRBG.

Hence discussed in triage to move this to backlog.

CC @ballifatih

**Comment by ballifatih** [comment]
I agree that we can backlog this for the moment, because we are pursuing SW-based DRBG as an alternative. For the future releases, I think there is a benefit in visiting this interface and providing more CSR space so that SW can input customization string along with 384-bit seed material.

**Comment by vogelpi** [comment]
Our CSRNG implements CTR_DRBG without derivation function (See "When a Derivation Function is Not Used" in the NIST spec). This variant doesn't support a nonce. Min/max entropy input lengths, max personalization string length and max additional input length are all seedlen which is 384 bits in our case.

The implementation supports a _personalization string_ for the Instantiate command, and _additional data_ for the Reseed and Generate commands, and _provided data_ for the Update command. See Section 10.2.1 in [NIST SP 800-90A](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf) and https://opentitan.org/book/hw/ip/csrng/doc/theory_of_operation.html#command-description for details.

How software can use this is slightly depends on the application interfaces:
- SW App: Software has full control over the commands and can provide _personalization string_, _addititonal data_ and _provided data_ all via the command (see command length or clen parameter).
- EDNs:
  - The Instantiate command is always software defined, so software can provide a _personalization string_
  - Auto Mode:
    - The Reseed and Generate commands are defined in the FIFOs in auto mode and software can provide _additional data_ via these FIFOs. These FIFOs will then be replayed, i.e., multiple Reseed and Generate commands will use the same _additional data_
    - There is no explicit Update command and no _provided data_.
  - SW Port Mode:
    - In SW Port Mode, also the Reseed and Generate commands are software defined. Software can use the `edn_cmd_req_done` interrupt (see https://opentitan.org/book/hw/ip/edn/doc/interfaces.html#interrupts) to find out when a potentially long Reseed or Generate command is done and then provide the next one. This way, all commands can get individual _personalization string_, _additional data_ or _provided data_ strings if needed. I acknowledge that this is probably not obvious from the doc as there is no good documentation on the SW Port Mode. The best I could find is here: https://opentitan.org/book/hw/ip/edn/doc/theory_of_operation.html#interaction-with-csrng-application-interface-ports

So for Earl Grey 2, I am moving this to Backlog (changing the implementation to use a derivation functions is infeasible from a timeline perspective). If this needs to be considered for future revisions, then we need to understand why the current implementation described above is not sufficient.
